j5/backends/hardware/sr/v4/motor_board.py

```diff
@@ -0,0 +1,190 @@
+from j5.boards import Board
+from j5.boards.sr.v4.motor_board import MotorBoard
+from j5.components.motor import MotorInterface, MotorSpecialState, MotorState
+
+CMD_RESET = 0
+CMD_VERSION = 1
+CMD_MOTOR = [2, 3]
+CMD_BOOTLOADER = 4
+
+SPEED_COAST = 1
+SPEED_BRAKE = 2
+
+RT = TypeVar("RT")  # pragma: nocover
+
+
+def handle_serial_error(func: Callable[..., RT]) -> Callable[..., RT]:  # type: ignore
+    """
+    Wrap functions that use the serial port, and rethrow the errors.
+
+    This is a decorator that should be used to wrap any functions that call the serial
+    interface. It will catch and rethrow the errors as a CommunicationError, so that it
+    is more explicit what is going wrong.
+    """
+    @wraps(func)
+    def catch_exceptions(*args, **kwargs):  # type: ignore
+        try:
+            return func(*args, **kwargs)
+        except SerialTimeoutException as e:
+            raise CommunicationError(f"Serial Timeout Error: {e}")
+        except SerialException as e:
+            raise CommunicationError(f"Serial Error: {e}")
+    return catch_exceptions
+
+
+class SRV4MotorBoardHardwareBackend(
+    MotorInterface,
+    Backend,
+):
+    """The hardware implementation of the SR v4 motor board."""
+
+    environment = HardwareEnvironment
+    board = MotorBoard
+
+    @classmethod
+    def discover(
+            cls,
+            find: Callable = comports,
+            serial_class: Type[Serial] = Serial,
+    ) -> List[Board]:
+        """Discover all connected motor boards."""
+        # Find all serial ports.
+        ports: List[ListPortInfo] = find()
+
+        # Filter down to just motor boards.
+        ports = list(filter(lambda x: x.vid == 0x403, ports))  # FTDI, Ltd
+        ports = list(filter(lambda x: x.pid == 0x6001, ports))  # FT232 Serial (UART) IC
+        ports = list(filter(lambda x: x.manufacturer == "Student Robotics", ports))
+        ports = list(filter(lambda x: x.product == "MCV4B", ports))
+
+        # Get a list of boards from the ports.
+        boards: List[Board] = []
+        for port in ports:
+            boards.append(
+                MotorBoard(
+                    port.serial_number,
+                    SRV4MotorBoardHardwareBackend(port.device, serial_class),
+                ),
+            )
+
+        return boards
+
+    @handle_serial_error
+    def __init__(self, serial_port: str, serial_class: Type[Serial] = Serial) -> None:
+        # Initialise our stored values for the state.
+        self._state: List[MotorState] = [
+            MotorSpecialState.BRAKE
+            for _ in range(0, 2)
+        ]
+
+        self._serial = serial_class(
+            port=serial_port,
+            baudrate=1000000,
+            timeout=0.25,
+        )
+
+        # Check we have the correct firmware version.
+        version = self.get_firmware_version()
+        if version != "3":
+            raise CommunicationError(
+                f"Unexpected firmware version: {version}, expected: 3.",
```

Comment:
The comparison above is for a string, this message suggests int?

Reply:
So we decided a while back that a `firmware_version` is a `str`, would changing the error to `f"Unexpected firmware version: {version}, expected: \"3\"."` help?

Reply:
It would, would the user be able to manipulate firmware_version by any chance? Wondering if there is any point to this

Reply:
`firmware_version` should be a read-only attribute

Reply:
4b2cfd1